File: integrations/alga-agent-sdk-py/alga_agent_sdk/commands.py
from __future__ import annotations

from typing import Any, Optional

from pydantic import BaseModel, ConfigDict


class InvestigationCommand(BaseModel):
    model_config = ConfigDict(extra="allow")

    op: str
    chat_id: Optional[str] = None
    fingerprint: Optional[str] = None
    reason: Optional[str] = None
    note: Optional[str] = None
    priority: Optional[str] = None
    severity: Optional[str] = None
    title: Optional[str] = None
    root_cause: Optional[str] = None
    resolution: Optional[str] = None
    summary: Optional[str] = None
    impact_assessment: Optional[str] = None
    actions_taken: Optional[str] = None
    eta: Optional[str] = None
    triage_result_id: Optional[str] = None
    agreed: Optional[bool] = None
    correct_decision: Optional[str] = None
    correct_severity: Optional[str] = None
    target_agent_id: Optional[str] = None
    role_type: Optional[str] = None
    user_id: Optional[str] = None
    agent_token_id: Optional[str] = None
    scope_description: Optional[str] = None
    incident_number: Optional[int] = None
    message: Optional[str] = None
    audience: Optional[str] = None
    urgency: Optional[str] = None
    status_level: Optional[str] = None
    source_coordination_message_id: Optional[str] = None
    internal: Optional[bool] = None


def resolve_alert(fingerprint: str) -> InvestigationCommand:
    return InvestigationCommand(op="resolve_alert", fingerprint=fingerprint)


def reopen_alert(fingerprint: str) -> InvestigationCommand:
    return InvestigationCommand(op="reopen_alert", fingerprint=fingerprint)


def set_outcome(
    root_cause: str | None = None,
    resolution: str | None = None,
) -> InvestigationCommand:
    return InvestigationCommand(op="set_outcome", root_cause=root_cause, resolution=resolution)


def cancel_investigation(reason: str | None = None) -> InvestigationCommand:
    return InvestigationCommand(op="cancel_investigation", reason=reason)


def pause_investigation(reason: str | None = None) -> InvestigationCommand:
    return InvestigationCommand(op="pause_investigation", reason=reason)


def triage_feedback(
    triage_result_id: str,
    agreed: bool = True,
    correct_decision: str | None = None,
    correct_severity: str | None = None,
    note: str | None = None,
) -> InvestigationCommand:
    return InvestigationCommand(
        op="triage_feedback",
        triage_result_id=triage_result_id,
        agreed=agreed,
        correct_decision=correct_decision,
        correct_severity=correct_severity,
        note=note,
    )


def assign_investigation(target_agent_id: str) -> InvestigationCommand:
    return InvestigationCommand(op="assign_investigation", target_agent_id=target_agent_id)


def promote_to_incident(
    title: str | None = None,
    severity: str | None = None,
    priority: str | None = None,
) -> InvestigationCommand:
    return InvestigationCommand(
        op="promote_to_incident", title=title, severity=severity, priority=priority
    )


def set_incident_priority(incident_number: int, priority: str) -> InvestigationCommand:
    return InvestigationCommand(
        op="set_incident_priority", incident_number=incident_number, priority=priority
    )


def set_incident_severity(incident_number: int, severity: str) -> InvestigationCommand:
    return InvestigationCommand(
        op="set_incident_severity", incident_number=incident_number, severity=severity
    )


def trigger_escalation(incident_number: int) -> InvestigationCommand:
    return InvestigationCommand(op="trigger_escalation", incident_number=incident_number)


def mitigate_incident(incident_number: int, reason: str | None = None) -> InvestigationCommand:
    return InvestigationCommand(
        op="mitigate_incident", incident_number=incident_number, reason=reason
    )


def resolve_incident(incident_number: int, reason: str | None = None) -> InvestigationCommand:
    return InvestigationCommand(
        op="resolve_incident", incident_number=incident_number, reason=reason
    )


def begin_triage(incident_number: int) -> InvestigationCommand:
    return InvestigationCommand(op="begin_triage", incident_number=incident_number)


def promote_incident(incident_number: int) -> InvestigationCommand:
    return InvestigationCommand(op="promote_incident", incident_number=incident_number)


def assign_incident_role_to_user(
    incident_number: int,
    role_type: str,
    user_id: str,
    scope_description: str | None = None,
) -> InvestigationCommand:
    return InvestigationCommand(
        op="assign_incident_role",
        incident_number=incident_number,
        role_type=role_type,
        user_id=user_id,
        scope_description=scope_description,
    )


def assign_incident_role_to_agent(
    incident_number: int,
    role_type: str,
    agent_token_id: str,
    scope_description: str | None = None,
) -> InvestigationCommand:
    return InvestigationCommand(
        op="assign_incident_role",
        incident_number=incident_number,
        role_type=role_type,
        agent_token_id=agent_token_id,
        scope_description=scope_description,
    )


def post_handoff(
    incident_number: int, message: str, audience: str, urgency: str
) -> InvestigationCommand:
    return InvestigationCommand(
        op="post_handoff",
        incident_number=incident_number,
        message=message,
        audience=audience,
        urgency=urgency,
    )


def publish_status_update(
    incident_number: int, message: str, status_level: str
) -> InvestigationCommand:
    return InvestigationCommand(
        op="publish_status_update",
        incident_number=incident_number,
        message=message,
        status_level=status_level,
    )


def set_incident_resolution_docs(
    incident_number: int,
    summary: str | None = None,
    impact_assessment: str | None = None,
    actions_taken: str | None = None,
    root_cause: str | None = None,
    resolution: str | None = None,
) -> InvestigationCommand:
    return InvestigationCommand(
        op="set_incident_resolution_docs",
        incident_number=incident_number,
        summary=summary,
        impact_assessment=impact_assessment,
        actions_taken=actions_taken,
        root_cause=root_cause,
        resolution=resolution,
    )

File: integrations/alga-agent-sdk-rs/src/commands.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct InvestigationCommand {
    pub op: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub chat_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub fingerprint: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub note: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub priority: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub severity: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub root_cause: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub resolution: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub impact_assessment: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub actions_taken: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub eta: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub triage_result_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agreed: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correct_decision: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub correct_severity: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_agent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub agent_token_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub scope_description: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub incident_number: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub audience: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub urgency: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status_level: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub source_coordination_message_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub internal: Option<bool>,
}

impl InvestigationCommand {
    fn new(op: &str) -> Self {
        Self {
            op: op.to_string(),
            ..Default::default()
        }
    }
}

fn opt(val: &str) -> Option<String> {
    if val.is_empty() {
        None
    } else {
        Some(val.to_string())
    }
}

pub fn resolve_alert(fingerprint: &str) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("resolve_alert");
    cmd.fingerprint = Some(fingerprint.to_string());
    cmd
}

pub fn reopen_alert(fingerprint: &str) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("reopen_alert");
    cmd.fingerprint = Some(fingerprint.to_string());
    cmd
}

pub fn set_outcome(
    root_cause: Option<&str>,
    resolution: Option<&str>,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("set_outcome");
    cmd.root_cause = root_cause.map(|s| s.to_string());
    cmd.resolution = resolution.map(|s| s.to_string());
    cmd
}

pub fn cancel_investigation(reason: &str) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("cancel_investigation");
    cmd.reason = opt(reason);
    cmd
}

pub fn pause_investigation(reason: &str) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("pause_investigation");
    cmd.reason = opt(reason);
    cmd
}

pub fn triage_feedback(
    triage_result_id: &str,
    agreed: bool,
    correct_decision: &str,
    correct_severity: &str,
    note: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("triage_feedback");
    cmd.triage_result_id = Some(triage_result_id.to_string());
    cmd.agreed = if agreed { Some(true) } else { None };
    cmd.correct_decision = opt(correct_decision);
    cmd.correct_severity = opt(correct_severity);
    cmd.note = opt(note);
    cmd
}

pub fn assign_investigation(target_agent_id: &str) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("assign_investigation");
    cmd.target_agent_id = Some(target_agent_id.to_string());
    cmd
}

pub fn promote_to_incident(
    title: &str,
    severity: &str,
    priority: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("promote_to_incident");
    cmd.title = opt(title);
    cmd.severity = opt(severity);
    cmd.priority = opt(priority);
    cmd
}

pub fn set_incident_priority(
    incident_number: i64,
    priority: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("set_incident_priority");
    cmd.incident_number = Some(incident_number);
    cmd.priority = Some(priority.to_string());
    cmd
}

pub fn set_incident_severity(
    incident_number: i64,
    severity: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("set_incident_severity");
    cmd.incident_number = Some(incident_number);
    cmd.severity = Some(severity.to_string());
    cmd
}

pub fn trigger_escalation(incident_number: i64) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("trigger_escalation");
    cmd.incident_number = Some(incident_number);
    cmd
}

pub fn mitigate_incident(
    incident_number: i64,
    reason: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("mitigate_incident");
    cmd.incident_number = Some(incident_number);
    cmd.reason = opt(reason);
    cmd
}

pub fn resolve_incident(
    incident_number: i64,
    reason: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("resolve_incident");
    cmd.incident_number = Some(incident_number);
    cmd.reason = opt(reason);
    cmd
}

pub fn begin_triage(incident_number: i64) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("begin_triage");
    cmd.incident_number = Some(incident_number);
    cmd
}

pub fn promote_incident(incident_number: i64) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("promote_incident");
    cmd.incident_number = Some(incident_number);
    cmd
}

pub fn assign_incident_role_to_user(
    incident_number: i64,
    role_type: &str,
    user_id: &str,
    scope_description: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("assign_incident_role");
    cmd.incident_number = Some(incident_number);
    cmd.role_type = Some(role_type.to_string());
    cmd.user_id = Some(user_id.to_string());
    if !scope_description.is_empty() {
        cmd.scope_description = Some(scope_description.to_string());
    }
    cmd
}

pub fn assign_incident_role_to_agent(
    incident_number: i64,
    role_type: &str,
    agent_token_id: &str,
    scope_description: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("assign_incident_role");
    cmd.incident_number = Some(incident_number);
    cmd.role_type = Some(role_type.to_string());
    cmd.agent_token_id = Some(agent_token_id.to_string());
    if !scope_description.is_empty() {
        cmd.scope_description = Some(scope_description.to_string());
    }
    cmd
}

pub fn post_handoff(
    incident_number: i64,
    message: &str,
    audience: &str,
    urgency: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("post_handoff");
    cmd.incident_number = Some(incident_number);
    cmd.message = Some(message.to_string());
    cmd.audience = Some(audience.to_string());
    cmd.urgency = Some(urgency.to_string());
    cmd
}

pub fn publish_status_update(
    incident_number: i64,
    message: &str,
    status_level: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("publish_status_update");
    cmd.incident_number = Some(incident_number);
    cmd.message = Some(message.to_string());
    cmd.status_level = Some(status_level.to_string());
    cmd
}

pub fn set_incident_resolution_docs(
    incident_number: i64,
    summary: &str,
    impact_assessment: &str,
    actions_taken: &str,
    root_cause: &str,
    resolution: &str,
) -> InvestigationCommand {
    let mut cmd = InvestigationCommand::new("set_incident_resolution_docs");
    cmd.incident_number = Some(incident_number);
    cmd.summary = opt(summary);
    cmd.impact_assessment = opt(impact_assessment);
    cmd.actions_taken = opt(actions_taken);
    if !root_cause.is_empty() {
        cmd.root_cause = Some(root_cause.to_string());
    }
    if !resolution.is_empty() {
        cmd.resolution = Some(resolution.to_string());
    }
    cmd
}
